tools/download-s3.py

```diff
@@ -0,0 +1,154 @@
+dry_run = False
+
+
+class Artifact (object):
+    def __init__(self, arts, path, info=None):
+        self.path = path
+        # Remove unexpanded AppVeyor $(..) tokens from filename
+        self.fname = re.sub(r'\$\([^\)]+\)', '', os.path.basename(path))
+        self.lpath = os.path.join(arts.dlpath, self.fname)
+        self.info = info
+        self.arts = arts
+        arts.artifacts.append(self)
+
+    def __repr__(self):
+        return self.path
+
+    def download(self, dirpath):
+        """ Download artifact from S3 and store in dirpath directory.
+            If the artifact is already downloaded nothing is done. """
+        if os.path.isfile(self.lpath) and os.path.getsize(self.lpath) > 0:
+            return
+        print('Downloading %s -> %s' % (self.path, self.lpath))
+        if dry_run:
+            return
+        self.arts.s3_bucket.download_file(self.path, self.lpath)
+
+
+class Artifacts (object):
+    def __init__(self, gitref, dlpath):
+        super(Artifacts, self).__init__()
+        self.gitref = gitref
+        self.artifacts = list()
+        # Download directory
+        self.dlpath = dlpath
+        if not os.path.isdir(self.dlpath):
+            if not dry_run:
+                os.makedirs(self.dlpath, 0o755)
+
+    def collect_single_s3(self, path):
+        """ Collect single S3 artifact
+         :param: path string: S3 path
+        """
+
+        # The S3 folder contains the tokens needed to perform
+        # matching of project, gitref, etc.
+        folder = os.path.dirname(path)
+
+        rinfo = re.findall(r'(?P<tag>[^-]+)-(?P<val>.*?)__', folder)
+        if rinfo is None or len(rinfo) == 0:
+            # print('Incorrect folder/file name format for %s' % folder)
+            return None
+
+        info = dict(rinfo)
+
+        # Ignore AppVeyor Debug builds
+        if info.get('bldtype', '').lower() == 'debug':
+            print('Ignoring debug artifact %s' % folder)
+            return None
+
+        tag = info.get('tag', None)
+        if tag is not None and (len(tag) == 0 or tag.startswith('$(')):
+            # AppVeyor doesn't substite $(APPVEYOR_REPO_TAG_NAME)
+            # with an empty value when not set, it leaves that token
+            # in the string - so translate that to no tag.
+            tag = None
+
+        sha = info.get('sha', None)
+
+        # Match tag or sha to gitref
+        if (tag is not None and tag == self.gitref) or (sha is not None and sha.startswith(self.gitref)):
+            return Artifact(self, path, info)
+
+        return None
+
+    def collect_s3(self):
+        """ Collect and download build-artifacts from S3 based on git reference """
+        print('Collecting artifacts matching tag/sha %s from S3 bucket %s' % (self.gitref, s3_bucket))
+        self.s3 = boto3.resource('s3')
+        self.s3_bucket = self.s3.Bucket(s3_bucket)
+        self.s3.meta.client.head_bucket(Bucket=s3_bucket)
+        for key in self.s3_bucket.objects.all():
```

Comment:
not a huge deal right now, but this will get increasingly slow over time as S3 listing isn't always particularly fast and it seems like this bucket will get uploads for all our CI builds as well. Would be good to think about if there is a layout of the paths that will allow you to restrict the set of objects that need to be listed.

Reply:
It will be limited to just tagged builds, but with 3-4 client repositories your point still stands (and it is already kind of slow..).

From a matching perspective we look for two things: the project (confluent-kafka-python, which is already a top-level directory) and the git ref (tag or sha) which is currently included in the sub-directory.
Multiple CI services may provide artifacts for the same git ref, and even multiple artifacts from the same service (e.g., travis-osx, travis-linux, windows-x64, windows-x86, ..).

It is not possible to conditionally change the deploy path based on whether a tag is present or not, so the best we can do is:
`confluent-kafka-python/<TAG>-<SHA>/<ci>-<variant>...`

And since we might not be confident that we know both the sha and tag at the same time we would still need to enumerate on confluent-kafka-python/*.

We could use just the tag, `confluent-kafka-python/<TAG>/`, but there is no guarantee that it will be unique if the tag is later moved and artifacts rebuilt, which would result in conflicting artifacts under that tag using different SHAs.

I don't know what's best.

Reply:
Kept the current format for now.

Reply:
Isn't the sha the thing you ultimately care about? Why doesn't organizing it as `confluent-kafka-python/<sha>/<anything else you want>` reduce it to basically the minimal set of files you'd need to scan through? Seems like in this script you can just convert tag -> sha if needed.